spring-pulsar/src/main/java/org/springframework/pulsar/core/PulsarTemplate.java

```diff
@@ -177,13 +180,22 @@ private MessageId doSend(@Nullable String topic, T message, @Nullable Schema<T>
 		}
 	}
 
-	private CompletableFuture<MessageId> doSendAsync(@Nullable String topic, T message, @Nullable Schema<T> schema,
-			@Nullable Collection<String> encryptionKeys,
+	private CompletableFuture<MessageId> doSendAsync(@Nullable String topic, @Nullable T message,
+			@Nullable Schema<T> schema, @Nullable Collection<String> encryptionKeys,
 			@Nullable TypedMessageBuilderCustomizer<T> typedMessageBuilderCustomizer,
 			@Nullable ProducerBuilderCustomizer<T> producerCustomizer) throws PulsarClientException {
 		String defaultTopic = Objects.toString(this.producerFactory.getProducerConfig().get("topicName"), null);
-		String topicName = this.topicResolver.resolveTopic(topic, message, () -> defaultTopic).orElseThrow(
-				() -> new IllegalArgumentException("Topic must be specified when no default topic is configured"));
+		String topicName = this.topicResolver.resolveTopic(topic, message, () -> defaultTopic).orElseThrow(() -> {
+			if (this.topicResolver.getClass().equals(DefaultTopicResolver.class)) {
```

Comment:
Yeh, I saw this before, it made me think of returning an `Either` (result/error) **OR** adding a `forceResolveTopic` (or force param or something like that) to signal to throw exception if not resolved. As the `if message type .. use this ex else use this ex" is definitely an impl. concern of the DTR.  

On one hand I am ok w/ just keeping this simple and using a generic msg altogether.
On the other hand, the "when null" could be helpful as that is the tombstone case and it may give users clue as to what the issue is.

I'm fine w/ whatever approach you end up going w/. If you are good w/ this one - simply resolve the comment. Thx

Reply:
**OR** we could change the TR signature to return ~ `ResolveTopicResult` which could contain an optional resolved topic OR a reason why it could not be resolved. That would remove the need for special message casing here and allow custom impls ability to tell users about why their custom resolver did not work. Wdyt?

Reply:
Good idea. I've added a `Resolved<T>` wrapper for this. PTAL